Mekhanik evgenii/fix 1346 1

EvgeniiMekhanik opened the pull request.

EvgeniiMekhanik committed:
Implement new config option `client_mem`.

client_mem <soft_limit> <hard_limit> - controls haw many
memory is used to store unanswered client requests and
requests with linked responses which can not be forwarded
to a client.

EvgeniiMekhanik committed:
Codestyle fixes

EvgeniiMekhanik committed:
Change API of some functions.

To track socket memory we should pass TfwHttpMsg * not
TfwMsgIter * to most of http_nsg_* functions, because
TfwHttpMsg has a pointer to connection and socket.

EvgeniiMekhanik committed:
Adjust memory used by Tempesta FW.

In task #498 we decide to use `client_mem` option to limit
count of memory used by client. This commit is a part of
this task - now Tempesta FW uses `sk->sk_rmem_alloc` to
adjust memory used by Tempesta FW for this client connection.

EvgeniiMekhanik committed:
Adjust memory used by Tempesta FW.

In task #498 we decide to use `client_mem` option to limit
count of memory used by client. This commit is a part of
this task and the next step of implementaion. Previosly
Tempesta FW uses `sk->sk_rmem_alloc` to adjust memory
used by Tempesta FW for this client connection, now we
adjust memory for the whole TfwClient, because the can
be a lot of connection for one client and for all other
cases we use limitation for TfwClient and block it if
necessary.

EvgeniiMekhanik committed:
Drop connetion with TCP RST if client mem is exceeded

If administrator specify `client_mem` and the memory
used by all connection of current client exceeded this
value Tempesta FW drops connection and block client
by ip if `ip_block on;` is specified.

EvgeniiMekhanik committed:
Inplement TfwClient refcounting for memory adjustment

Previosuly we get connection when we adjust memory
for skb, but it leads to several problems:
- we can't adjust memory for skb before tls decryption,
because skb from `tls->io_in.skb_list` are freed
during connection released (but connection will be
never released if we increment it's reference counter
for these skbs).
- We have the same problems for skbs, which are wait
for appropriate tcp window to be pushed in socket
write queue.
Now we increment/decrement reference counter for
TfwClient and adjust skb memory for requests before
tls decryption.

EvgeniiMekhanik committed:
Adjust tcp send window for all sending data.

Previously we adjust tcp send window only for http2
connection and only during making HEADER or DATA frames,
but if we want to control client memory usage we should
do it for all type of sending data. (We orphane skb and
decrease memory usage when we pass skb to the socket
write queue, so we we don't adjust tcp send window we
push a lot of skbs in socket write queue and don't
adjust it's memory).

EvgeniiMekhanik committed:
Fix accroding review and after rebase

- remove `client_get_light/client_put_light` functions,
  because after removing lock from `client` structure
  we don't need these functions at all.
- Adjust memory usage of skb in `skb->cb`. Usually it
  is equal to `skb->truesize, but for some cases (
  skb which was created by `pskb_copy_for_clone` for
  example it is different).

const-t commented on fw/sock.c:
```diff
@@ -421,16 +428,21 @@ ss_skb_tcp_entail(struct sock *sk, struct sk_buff *skb, unsigned int mark,
 	int r;
 
-	while ((skb = ss_skb_dequeue(skb_head))) {
+	while ((*snd_wnd = tfw_tcp_calc_snd_wnd(sk, mss_now))) {
```
We calculate window here, but we don't check it in the loop. I mean we can put into a `write_queue` skb with len 200 when `snd_wnd` is 1.

const-t commented on fw/sock.c:
```diff
@@ -397,40 +404,100 @@ ss_forced_mem_schedule(struct sock *sk, int size)
 }
 
+static bool
+ss_skb_can_collapce(struct sk_buff *to, struct sk_buff *from,
```
```suggestion
ss_skb_can_collapse(struct sk_buff *to, struct sk_buff *from,
```

const-t reviewed (approved):
LGTM

krizhanovsky commented on fw/http2.c:
```diff
@@ -647,7 +647,7 @@ tfw_h2_stream_xmit_prepare_resp(TfwStream *stream)
 	BUG_ON(!resp || resp->msg.skb_head || !resp->req
 	       || !resp->req->conn || !stream->xmit.skb_head);
```
It seems comment https://github.com/tempesta-tech/tempesta/issues/1346#issuecomment-1611029794 isn't addressed. Unfortunately the comment doesn't say how exactly `tfw_h2_closed_streams_shrink()` should be changed and why, so I'm not sure if anyone can remember what the comment is about.

krizhanovsky commented on fw/ss_skb.c:
```diff
@@ -1336,10 +1343,16 @@ ss_skb_init_for_xmit(struct sk_buff *skb)
+
+	/*
+	 * dev is used to save connection for memory accounting
+	 * clear it before pass skb to the kernel.
```
I didn't find where we actually store connection in `skb->dev`

krizhanovsky commented on fw/ss_skb.c:
```diff
@@ -1706,10 +1719,84 @@ int
+		mem = TFW_SKB_CB(skb)->mem;
+		old_truesize = skb->truesize;
+		skb_orphan(skb);
+	}
```
Do we really need to play with the skb owner? This looks risky and basically `client_mem` is our own accounting. Probably we can account expanded head in parallel with the socket/skb accounting or just do not care about `pskb_expand_head()`.

This might open a DoS attack vector, but this is about _mitigation_, so in general we can afford some accounting inaccuracy to get faster and more reliable code.

krizhanovsky commented on fw/ss_skb.c:
```diff
@@ -1335,10 +1339,16 @@ ss_skb_init_for_xmit(struct sk_buff *skb)
 	__u8 pfmemalloc = skb->pfmemalloc;
 
-	WARN_ON_ONCE(skb->sk);
+	skb_orphan(skb);
```
Why don't we make a pointer to a client accounting in `skb->cb` instead of to play with `skb_orphan()`? I'd prefer to avoid this since we can get plenty of crashes in this patch or in later kernel version migrations due to breaking kernel logic about orphaned skbs.

krizhanovsky commented on fw/connection.c:
```diff
@@ -231,3 +232,102 @@ tfw_connection_unlink_to_sk(TfwConn *conn)
+int
+tfw_connection_push(TfwConn *conn, unsigned int mss_now)
+{
+	struct sock *sk = conn->sk;
```
`conn` dereferencing must be done after `BUG_ON` at the below

krizhanovsky commented on fw/connection.c:
```diff
@@ -231,3 +232,102 @@ tfw_connection_unlink_to_sk(TfwConn *conn)
+         */
+	BUG_ON(!conn);
+
+	BUG_ON(SS_CONN_TYPE(sk) & Conn_Closing);
```
Probably this can be just a warning. Also probably we don't need an empty line before this assertion

krizhanovsky commented on fw/ss_skb.c:
```diff
@@ -1706,10 +1719,84 @@ int
+		BUG_ON(skb->sk);
+		WARN_ON(TFW_SKB_CB(skb)->mem != 0);
+
+		tfw_client_get((TfwClient *)owner);
```
This is an additional reference count for the TDB entry - were is corresponding decerment of the reference count?

krizhanovsky commented on fw/ss_skb.c:
```diff
@@ -1706,10 +1719,84 @@ int
+		BUG_ON(TFW_SKB_CB(skb)->mem < 0);
+		tfw_client_adjust_mem(cli, delta);
+	}
+}
```
At the moment only this function accounts `TfwClient->mem` and the function accounts `skb` data only. I believe we also need to account HTTP level data as well. For example, a single skb may produce many HTTP messages and each of the message may unpack a lot of data from HPACK.

Not necessary to account everything supper accurately, but the main data consumers must be accounted.

krizhanovsky reviewed (changes requested):
Quite a big and sophisticated change, definitely requiring some discussion. Let's discuss the PR on the next meeting. Besides absent logic and bugs, I see following topics to consider/discuss:
1. how the soft limit of `client_mem` is supposed to be implemented? Shouldn't we revert the `sk->sk_rmem_alloc` patch? I did like the approach since it naturally implements the soft limit, but it's unclear how to implement it now
2. It seems for me that there is no sense to play with `skb_orphan()` if we account everything on application level (no `sk->sk_rmem_alloc` and only `TfwClient` accounting)It seems  `skb_orphan()` is required only for more accurate accounting for `pskb_expand_head()`, but does it really impacts much into overall memory consumption?
3. I believe we also should account HTTP level memory
4. There are test pull requests - https://github.com/tempesta-tech/tempesta-test/pull/615 and https://github.com/tempesta-tech/tempesta-test/pull/618 - but they both are in draft. They must be finished along with this PR and they do need to test accounting accuracy and #1715 scenario - it must be fixed with this PR.

Also it seems there are 2 irrelevant commits for the #1346 and #1715 logic and it'd be better to publish such commits separately to make PRs review and merge faster.

Please make all upcoming changes in separate commits and please write good commit messages even for commits supposed to be squashed.

EvgeniiMekhanik committed:
Fix according review

Do not use `skb->sk` and `skb->destructor` to check
memory used by skb, use `skb->cb` for this purposes.
- Implement our own version of `skb_orphan` with
name `ss_skb_orphan` which is called when skb is freed
in Tempesta FW code our just before pushing skb to
socket write queue.
- Implement wrappers over `__kfree_skb` and `kfree_skb`
  where we call `ss_skb_orphan` before free skb.
- Check that skb is pushed to socket write queue, using
  new ipmlemented function `skb_tfw_is_in_socket_write_queue`
  from linux kernel, to skip adjusting memory used be skb,
  when it belongs to kernel (when `ss_skb_*` functions
  called from `tls_encrypt`).

EvgeniiMekhanik committed:
Remove extra connections callbacks

- Usually we use callbacks which are set in `skb->cb`
  for different purposes. So remove to callbacks, which
  was added in previous patches and use callbacks saved
  in `skb->cb`.

EvgeniiMekhanik committed:
Adjust http level memory.

- Since we use pool for http memory allocation, change
  api of all `tfw_pool_*` functions to pass `TfwClient`
  and accounting memory in this structure.
- Remove `TfwClient` refcounter (it not used, can be done
  in previous commits).
- Fix unit tests to check memory accounting, cleanup memory
  after each test, to check that client memory is equal to
  zero after test.

const-t commented on fw/pool.h:
```diff
@@ -53,18 +53,20 @@ typedef struct tfw_pool_chunk_t {
  */
 typedef struct {
 	TfwPoolChunk	*curr;
+	void		*owner;
```
Why not forward-declare `TfwClient` struct?

EvgeniiMekhanik committed:
Use per cpu counter instead of atomic

A big performance degradation was found after this patch.
During investigation it was found that the problem is in
usage atomic counter for client mem accounting. Usage per_cpu
array instead of atomic counter fix a performance issue.

EvgeniiMekhanik commented:
Big responses
master:
finished in 40.06s, 64391.32 req/s, 6.12GB/s
2e94835a50ff3f0f0c4b7fc08887e32e49c7f0fa:
finished in 40.10s, 63755.38 req/s, 6.11GB/s - no client mem option, just accounting

const-t commented on fw/client.c:
```diff
@@ -182,16 +252,39 @@ tfw_client_addr_eq(TdbRec *rec, void *data)
+tfw_client_free(TdbRec *rec)
+{
+        /* Stats should be updated(dec/inc) only for complete records. */
+        if (!tdb_entry_is_complete(rec))
```
I believe we don't need this condition for clients, we allocate per-cpu memory even for not complete records. At this moment we are safe, however if in future we fail before set record as complete we will have memleak.

const-t commented on fw/cache.c:
Here we add client mem for plain http1 as well as other protocols, however for plain http1 memory will not be allocated and `ss_skb_to_sgvec_with_new_pages()`will not be called. I'm not ask to change this behavior it is up to you, but I ask to write a comment to describe this.

const-t commented on fw/sock.c:
```diff
@@ -681,6 +668,10 @@ ss_send(struct sock *sk, struct sk_buff **skb_head, int flags)
+			memset(twin_skb->cb, 0, sizeof(twin_skb->cb));
+			ss_skb_set_owner(twin_skb, ss_skb_dflt_destructor,
+					 TFW_SKB_CB(skb)->opaque_data,
+					 skb_headlen(skb));
```
```suggestion
					 skb_headlen(twin_skb));
```

EvgeniiMekhanik committed:
Fix removing still used clients.

Previously we remove client entry from TDB if there
is no entry in `client_lru.free_list` and new client
is allocated, even if such removed client still have
any active connections. There is a BUG in such strategy -
if this removed client has hung connections, we can't
close and destroy them during Tempesta FW unloading,
because we close and destroy connections during iteration
through active clients (`tfw_client_for_each`).
In new strategy we change logic in `tdb_htrie_put_rec`.
We add pointer to the bucket in the record structure. When we
remove record we zeroed this pointer. If record reference counter
became equal to zero, but bucket pointer is still not NULL (record
was not removed) we remove such record from the bucket using
this pointer. For clients we just use tfw_client_put, without
record removing, when  client reference counter became equal
to zero client record will be removed from bucket and freed.

EvgeniiMekhanik committed:
Improve performance and make fixes

We can't call tfw_client_get/put on each allocated
or orphaned skb. (Or each pool creation/destroing).
Under pressure when we have a lot of cpus that lead
to atomic contention and bad performance degradation.
To fix this problem we implement special TfwClientMem
structure, with it's own reference accounting (using
struct percpu_ref!) and save in the client structure
point to it. We use percpu_ref_tryget/percpu_ref_put
during skb allocation/deallocation (it's very cheap).
When we destroy client we schedule work, call
`percpu_ref_kill_and_confirm` and wait until all skbs
will be orphaned.
Also make some fixes according review:
- Call `tfw_client_free` for incomplete records also.
- Implement `tfw_alloc_percpu_gfp` same as `alloc_percpu_gfp`
  but with error injection
- Fix memory accouting during copying skbs.

const-t commented on fw/client.c:
```diff
@@ -193,16 +188,69 @@ tfw_client_addr_eq(TdbRec *rec, void *data)
+{
+	TfwClientMem *cli_mem;
+
+	cli_mem = tfw_kmalloc(sizeof(TfwClientMem), GFP_ATOMIC);
```
It would be good to re-use this memory, good for memcache, but as I see, we can't use it, because during shutdown we may destroy the cache before freeing all objects.
 
Also this is untested place. We didn't do benchmarks with many clients so we can't say what overhead it introduces

EvgeniiMekhanik replied:
I thnik that new client initialization is much more rare than if we use atomic_inc?dec on each skb allocation for client reference counting

const-t replied:
> I thnik that new client initialization is much more rare than if we use atomic_inc?dec on each skb allocation for client reference counting

Indeed, however it depends on the size of lru, if we have a lot of unique clients we may have performance spikes. It is just warning from my side, I'm not sure that we need to change this especially without benchmark.

@krizhanovsky @EvgeniiMekhanik Anyway in this function `tfw_client_mem_alloc()` we have two allocations, one allocates `TfwClientMem` and the second inside `percpu_ref_init()`.

const-t replied:
@EvgeniiMekhanik I believe with wait queue, that was added in the last patch we can use memcache instead of kmalloc here.

EvgeniiMekhanik replied:
Yes fixed, moreover preallocated pool was implemeneted

EvgeniiMekhanik commented on fw/client.c:
```diff
@@ -193,16 +188,69 @@ tfw_client_addr_eq(TdbRec *rec, void *data)
+}
+
+static void
+cli_mem_release(struct percpu_ref *ref)
```
Yes fixed this and trace

const-t commented on fw/client.c:
```diff
@@ -65,6 +65,9 @@ static struct {
 
 static TDB *client_db;
 
+static atomic_t shutdown_pending = ATOMIC_INIT(0);
```
static variable doesn't require zero initialization.

EvgeniiMekhanik replied:
yes, but we do it in other places. see sock.c for example

const-t replied:
In some places we do, in some don't. In access_log.c we don't do.

keshonok replied:
Technically, you don't know if a macro results in a pure zero value. Moreover, you don't need to know that. You just need to use the macro where appropriate. I believe the use of the macro here is justified and correct.

EvgeniiMekhanik replied:
I agree with @keshonok. I think that it is good to use this macros everywhere. There is an api and we should use it

const-t commented on fw/client.c:
```diff
@@ -193,16 +194,73 @@ tfw_client_addr_eq(TdbRec *rec, void *data)
+	if (!cli_mem->mem)
+		goto free_cli_mem;
+
+	if (percpu_ref_init(&cli_mem->refcnt, cli_mem_release, 0, GFP_ATOMIC))
```
I can't see call `percpu_ref_exit()` we should do this, to free `data`. I believe we can call it from `release` callback.

EvgeniiMekhanik replied:
yes fixed

const-t commented on fw/client.c:
```diff
@@ -193,16 +194,73 @@ tfw_client_addr_eq(TdbRec *rec, void *data)
+{
+	TfwClientMem *cli_mem = container_of(ref, TfwClientMem, refcnt);
+
+	call_rcu(&cli_mem->rcu_head, __cli_mem_release);
```
It seems we are safe to call `__cli_mem_release()` right here, instead of `call_rcu()`

EvgeniiMekhanik replied:
Yes fixed. This is left over from my other idea)

const-t commented on fw/client.c:
```diff
@@ -65,6 +65,9 @@ static struct {
 
 static TDB *client_db;
 
```
@krizhanovsky @EvgeniiMekhanik All this memory counting introduce some overhead, maybe it worth not to allocate all per-cpu staff and so on if `client_mem` is zero? It anyway can't be reconfigured now.

EvgeniiMekhanik replied:
yes fixed

const-t commented on fw/pool.h:
```diff
@@ -53,18 +53,20 @@ typedef struct tfw_pool_chunk_t {
  */
 typedef struct {
 	TfwPoolChunk	*curr;
+	void		*owner;
```
Using forward declaration for `TfwClient` we can use type name here instead of `void *`, like we do for other structures, see `fw/http_types.h"

EvgeniiMekhanik replied:
Yes fixed

EvgeniiMekhanik committed:
Implement client mem allocation optimization

- Preallocate and initialize TfwClientMem structures
  according to client_cfg.lru_size. During TfwCLient
  structure allocation try to get TfwClientMem from
  preallocated pool. If this pool is empty alloc
  TfwClientMem from cache.
- Implement new fault injection alloc functions to
  cover new code.

1396:
small:
finished in 50.03s, 1294710.94 req/s, 1002.60MB/s
finished in 50.03s, 1287197.04 req/s, 998.01MB/s
large:
finished in 50.03s, 103497.42 req/s, 9.90GB/s
finished in 50.03s, 108665.42 req/s, 10.39GB/s
1396 with client_mem:
small:
finished in 50.03s, 1225390.98 req/s, 948.74MB/s
finished in 50.03s, 1223275.66 req/s, 947.72MB/s
large:
finished in 50.08s, 78906.58 req/s, 7.55GB/s
finished in 50.08s, 86201.98 req/s, 8.24GB/s
master:
small:
finished in 50.03s, 1294782.10 req/s, 1002.49MB/s
finished in 50.03s, 1294782.10 req/s, 1001.33MB/s
large:
finished in 50.04s, 98625.80 req/s, 9.43GB/s
finished in 50.04s, 97767.22 req/s, 9.35GB/s

EvgeniiMekhanik committed:
Fix BUG in case of allocation fail.

We should check that TlsMpiPool was allocated during
`ttls_mpool_exit`

EvgeniiMekhanik committed:
Fix according review

- Use cache for client mem allocations
- Use typedef for TfwClientMem structure and pass
  TfwClientMem * pointer instread of void * in all
  functions.
- Make client_mem option reconfigurable.

EvgeniiMekhanik committed:
Rework error code library.

- Remove SS_* errocodes, use only T_* error codes, since we already
  include "lib/log.h" in all files.
- Split enum with error codes to two enums (one for
  common error codes and one for internal).
- List all error codes from least crusial to most crusial
- Implement some useful functions.

EvgeniiMekhanik committed:
Improve performance during checking client memory

Don not check client memory consumption during parsing
requests/http2 frame processing, check it only at the
end of softirq rx path. Under load we can process little
bit more requests, before dropping connection and block
client but in this case we have no perfomance degradation:
1396 with memory check:
small:
finished in 50.03s, 1308389.70 req/s, 1013.29MB/s
finished in 50.03s, 1303456.74 req/s, 1010.62MB/s
finished in 50.03s, 1298829.02 req/s, 1007.03MB/s
large:
finished in 50.08s, 103259.90 req/s, 9.88GB/s
finished in 50.08s, 111437.44 req/s, 10.66GB/s
finished in 50.07s, 101983.44 req/s, 9.75GB/s
finished in 50.07s, 111777.22 req/s, 10.69GB/s
master:
small:
finished in 50.03s, 1330912.18 req/s, 1.01GB/s
finished in 50.03s, 1343435.90 req/s, 1.02GB/s
finished in 50.03s, 1344150.06 req/s, 1.02GB/s
large:
finished in 50.04s, 98945.70 req/s, 9.46GB/s
finished in 50.04s, 96055.18 req/s, 9.19GB/s
finished in 50.04s, 102439.50 req/s, 9.80GB/s

EvgeniiMekhanik committed:
Update copyright years

const-t commented on lib/log.h:
```diff
@@ -22,48 +22,100 @@
-	 * with TCP_FIN.
+	 * Common error codes high boundary. All common error codes
+	 * should be less then this boundary. Error codes should be listed
+	 * from least crusial to most crusial.
```
```suggestion
	 * from least crucial to most crucial.
```

const-t commented on lib/log.h:
```diff
@@ -22,48 +22,100 @@
+}
+
+static inline bool
+tfw_error_code_more_crusial(int err_code1, int err_code2)
```
```suggestion
tfw_error_code_more_crucial(int err_code1, int err_code2)
```

const-t commented on lib/log.h:
```diff
@@ -22,48 +22,100 @@
+}
+
+static inline bool
+tfw_error_code_more_crusial(int err_code1, int err_code2)
```
nit: From my point of view calling this function is more sophisticated than simple `err1 < err2`.

EvgeniiMekhanik replied:
According our private discussion we decide to use this function

const-t commented on lib/log.h:
```diff
@@ -22,48 +22,100 @@
+	 * should be less then this boundary. Error codes should be listed
+	 * from least crusial to most crusial.
+	 */
+	__T_INTERNAL_ERROR_CODE_MAX = -MAX_ERRNO + T_INTERNAL_ERROR_CODE_MAX,
```
Do we need `__T_INTERNAL_ERROR_CODE_MAX`? It is unclear for me what difference between common and internal error codes and why `T_BLOCK` is internal but `T_BLOCK_WITH_RST` is common.

const-t commented on lib/log.h:
```diff
@@ -22,48 +22,100 @@
+}
+
+static inline bool
+tfw_error_code_is_crusial(int err_code)
```
```suggestion
tfw_error_code_is_crucial(int err_code)
```

const-t commented on fw/client.c:
```diff
@@ -58,12 +59,175 @@ typedef struct {
+{
+	int r;
+
+	cli_mem->mem = tfw_alloc_percpu_gfp(long, GFP_KERNEL | __GFP_ZERO);
```
You calling `tfw_cli_mem_init()` from softIRQ as well, but you use only `GFP_KERNEL` flags. For `tfw_percpu_ref_init()` the same.

EvgeniiMekhanik replied:
yes fixed

const-t commented on fw/client.c:
```diff
@@ -58,12 +59,175 @@ typedef struct {
+		return -EINVAL; 
+
+	cli_mem_pool.order =
+		get_order(sizeof(TfwClientMem) * client_cfg.lru_size);
```
For large LRUs we potentially can fail and not allocate memory, I believe we must allocate not greater than `MAX_PAGE_ORDER`.

EvgeniiMekhanik replied:
yes fixed

const-t commented on fw/client.c:
```diff
@@ -159,16 +159,18 @@ tfw_cli_mem_kill_work_fn(struct work_struct *work)
 
 static inline int
-tfw_cli_mem_init(TfwClientMem *cli_mem)
+tfw_cli_mem_init(TfwClientMem *cli_mem, bool from_pool)
```
Maybe just pass flags would be better? But it is not important I'm ok with it

const-t reviewed (approved):
LGTM, however review from @krizhanovsky is required, especially to approve new approach with memory limit checking.

krizhanovsky commented on fw/sock_clnt.c:
```diff
@@ -922,6 +939,15 @@ static TfwCfgSpec tfw_sock_clnt_specs[] = {
 		.allow_reconfig = true,
 	},
+	{
+		.name = "client_mem",
```
@EvgeniiMekhanik @const-t yes, this definitely makes sense. The 1GB limit was just a workaround for #498 and it makes sense to completely remove it and replace with `client_mem`. Just make a temporary limit for the value as 1GB

krizhanovsky commented on fw/http.c:
```diff
@@ -6540,6 +6584,23 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, struct sk_buff *skb,
+				 conn->peer, skb->truesize);
+	}
+
+	r = frang_client_mem_limit((TfwCliConn *)conn, false);
```
@const-t is this comment still relevant? I see for example, that `ss_skb_adjust_data_len()` is called from`tfw_hpack_cache_decode_expand()` ->  `tfw_hpack_cache_decode_expand()` -> `tfw_http_msg_expand_data()`, i.e the current code does account HTTP responses

krizhanovsky commented on fw/http_frame.c:
```diff
@@ -2186,7 +2209,7 @@ do {									\
 #define FRAME_XMIT_FSM_NEXT(frame_length, state)			\
 do {									\
-	*snd_wnd -= frame_length;					\
+	*snd_wnd -= frame_length - FRAME_HEADER_SIZE;			\
```
" adjust FRAME_HEADER_SIZE during calculation send window during making frames." does not explain why the header frame size is subtracted. We do not have capacity to fix every small thing, but please **do put attention to commit messages and code commenting**. This is super important especially if such a sophisticated work as this PR and whole 1346

krizhanovsky commented on fw/connection.h:
```diff
@@ -573,6 +575,7 @@ tfw_connection_validate_cleanup(TfwConn *conn)
 	BUG_ON(!conn);
 	BUG_ON(!list_empty(&conn->list));
 	BUG_ON(conn->stream.msg);
+	BUG_ON(conn->write_queue);
```
The point was to comment this **in code, not in PR**

krizhanovsky commented on fw/client.c:
```diff
@@ -341,7 +342,6 @@ tfw_client_stop(void)
 	if (tfw_runstate_is_reconfig())
 		return;
 
-	tfw_client_free_lru();
```
It seems the content of commit https://github.com/tempesta-tech/tempesta/pull/2456/changes/a3ad1762d9fa13ca9d9ef9f2571b371c2bdd74bd and its message are about different things

krizhanovsky committed:
Adjust coding style and add a minor comment

krizhanovsky reviewed (approved):
LGTM for now. I listed all required to dos in https://github.com/tempesta-tech/tempesta/issues/1715#issuecomment-4313947521 and made a small commit to adjust coding style to not to pollute the comment. Let's address the comments in a new PR - it becomes to hard to review this PR - let's just merge it for now.

EvgeniiMekhanik committed:
Adjust frame size during `snd_wnd` calculation

Adjust FRAME_HEADER_SIZE during calculation of send window
during making frames. (There was a mistake with accuracy
of send window calculation, we don't take into account,
that each frame also contains frame header).
